core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -436,6 +439,14 @@ public void commit() {
     }
   }
 
+  private void cleanupAfterFailure(RuntimeException exception) {
+    if (!strictCleanup || exception instanceof CleanableFailure) {
+      Exceptions.suppressAndThrow(exception, this::cleanAll);
+    }
+
+    throw exception;
```

Comment:
This shouldn't throw an exception. The exception should be thrown from the calling code so that it is clear when reading that section how execution will happen. It's fine to pass it in here to determine whether to clean up though.

I'm not sure that this refactoring is worth the method, but it's up to you once this `throw` is moved back.

Reply:
Good point, it's not a good idea to have a helper method throw a runtime exception, it's less explicit to a reader since they'd have to actually step into the helper. I'll fix  this up